File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
class Solution {
    public int solution(int[][] maps) {
        int answer = 0;
        int n = maps.length;
        int m = maps[0].length;
        
        int[] dc = {1, 0, 0, -1};
        int[] dr = {0, 1, -1, 0};
        boolean[][] visited = new boolean[n][m];
        
        Deque<int[]> queue = new ArrayDeque<>();
        queue.add(new int[]{0, 0, 1});
        
        while(!queue.isEmpty()){
            int[] temp = queue.poll();
            int r = temp[0];
            int c = temp[1];
            int dist = temp[2];
            
            if(r==n-1 && c==m-1){
                return dist;
            }
            for(int i=0; i<4; i++){
                int nc = c+dc[i];
                int nr = r+dr[i];
                if(nc>=0 && nc<m && nr>=0 && nr<n && maps[nr][nc]==1){
                    if(!visited[nr][nc]){
                        visited[nr][nc]=true;
                        queue.add(new int[]{nr, nc, dist+1});
                    }
                }
            }
        }
        
        return -1;
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    var answer = 1;
    
    const dr = [1, -1, 0, 0];
    const dc = [0, 0, 1, -1];
    const visited = Array.from(Array(maps.length), ()=>Array(maps[0].length).fill(true));
    visited[0][0]=false;
    
    const q = [[0, 0, 1]]
    
    while(true){
        if(q.length===0){
            break;
        }
        const next = q.shift();
        const r = next[0];
        const c = next[1];
        const count = next[2];
        
        if(r==maps.length-1&&c==maps[0].length-1){
            return count;
        }
        
        for(let i=0; i<4; i++){
            const nr = r+dr[i];
            const nc = c+dc[i];
            if(nr>=0 && nr<maps.length && nc>=0 && nc<maps[0].length && maps[nr][nc]==1 && visited[nr][nc]){
                visited[nr][nc]=false;
                q.push([nr, nc, count+1])
            }
        }
    }
    
    return -1;
}